application/src/main/java/uk/nhs/hee/tis/revalidation/service/DoctorsForDBService.java

```diff
@@ -162,24 +160,52 @@ public void updateDoctorConnection(final ConnectionMessageDto message) {
     }
   }
 
+  /**
+   * Handle DoctorsForDbCollectedEvent as part of GMC Overnight Sync by disconnecting doctors for a
+   * given DBC that were not updated in this sync period.
+   *
+   * @param doctorsForDbCollectedEvent event that signifies that all doctors for a given DB have
+   *                                   been collected
+   */
+  public void handleDoctorsForDbCollectedEvent(
+      final DoctorsForDbCollectedEvent doctorsForDbCollectedEvent) {
+    final String designatedBodyCode = doctorsForDbCollectedEvent.designatedBodyCode();
+    final LocalDateTime requestDateTime = doctorsForDbCollectedEvent.requestDateTime();
+
+    //There are expected to be no more than 8,000 small records, well within limits on payload and transaction limits
+    doctorsRepository.saveAll(doctorsForDbCollectedEvent.doctors());
+    List<DoctorsForDB> staleDoctors = doctorsRepository.findByDesignatedBodyCodeAndGmcLastUpdatedDateTimeBefore(
+        designatedBodyCode, requestDateTime);
+    staleDoctors.forEach(d -> {
+      //To minimise issues of concurrency ensure we've got the latest since we first collected the batch
+      var doctorForDB = doctorsRepository.findById(d.getGmcReferenceNumber());
+      doctorForDB.ifPresentOrElse(
+          doctorsForDB -> {
+            if (!doctorsForDB.getDesignatedBodyCode().equals(designatedBodyCode)
+                || requestDateTime.isBefore(doctorsForDB.getGmcLastUpdatedDateTime())) {
```

Comment:
I'm drawing attention to this as I hadn't previously considered what should happen when at this point the designated body code has changed but the "new" last updated time is still earlier than this `requestDateTime`?

At first glance I though we should 'disconnect' them but there could have been a delay in that other change being applied which the designated body event wouldn't reflect.

Reply:
I think it's a good idea to have these logic, as this forEach may take some time, and if we disconnect them, the new connection data would be overwritten, which is not what we expect.